Library/Homebrew/formula.rb

```diff
@@ -1720,6 +1720,37 @@ def self.installed_with_alias_path(alias_path)
     installed.select { |f| f.installed_alias_path == alias_path }
   end
 
+  # An array of all installed {Formula} with {Cask} dependents.
+  # @private
+  def self.installed_formulae_with_cask_dependents
+    Cask::Caskroom.casks
+                  .flat_map { |cask| cask.depends_on[:formula] }
+                  .compact
+                  .map { |f| Formula[f] }
+                  .flat_map { |f| [f, *f.runtime_formula_dependencies].compact }
+  end
+
+  # An array of all installed {Formula} without {Formula} or {Cask} dependents
+  # that weren't installed on request.
+  # @private
+  def self.removable_formulae
+    formulae = installed
+    all_removable_formulae = T.let([], T::Array[Formula])
+
+    loop do
```

Comment:
I'm always a bit concerned if there's any way for these to run indefinitely. Thoughts?

Reply:
Not a super strong opinion, but I think the recursive formulation is a lot neater (and easier to understand). You also won't be able to recurse infinitely -- you'll get a `stack level too deep` error eventually.

That said, writing it recursively could be a bit (or a lot) slower. If it is, then we should definitely avoid the recursive form.